File: cpp/code/dataStructure/Trie.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

class Node {
public:
    bool isWord;
    string word;
    unordered_map<char, Node*> next;

    Node() {
        isWord = false;
    };
    ~Node() { next.clear(); }
};

class Trie {
private:
    Node* root;
public:
    void insert(string word) {
        Node* curNode = root;
        for (int i = 0; i < (int)word.size(); i++) {
            char c = word[i];
            if (curNode->next.count(c)) {
                curNode = curNode->next[c];
            }
            else {
                Node* n = new Node();
                n->word = c;
                curNode->next[c] = n;
                curNode = n;
            }
        }
        curNode->isWord = true;
    }

    bool search(string word) {
        Node* curNode = root;
        for (int i = 0; i < (int)word.size(); i++) {
            char c = word[i];
            curNode = curNode->next[c];
            if (!curNode) return false;
        }
        return curNode->isWord;
    }

    bool startWith(string prefix) {
        Node* curNode = root;
        for (int i = 0; i < prefix.size(); i++) {
            char c = prefix[i];
            curNode = curNode->next[c];
            if (!curNode) return false;
        }
        return true;
    }

    Trie() {
        root = new Node();
    }
    ~Trie() {
        delete root;
    }
};

int main() {
    Trie* trie = new Trie();
    trie->insert("app");
    trie->insert("apple");
    trie->insert("application");
    cout << trie->search("appl") << "\n";
    cout << trie->startWith("applicac") << "\n";

    return 0;
}

File: typescript/autoComplete/trie.ts
class TrieNode {
  public isWord: boolean;

  public word: string;

  public next: {
    [key: string]: TrieNode;
  };

  constructor() {
    this.isWord = false;

    this.next = {};

    this.word = "";
  }
}

class Trie {
  private root: TrieNode;

  constructor() {
    this.root = new TrieNode();
  }

  public insert = (word: string) => {
    let curNode: TrieNode = this.root;

    for (let i = 0; i < word.length; i++) {
      const c: string = word[i];
      if (!curNode.next[c]) {
        curNode.next[c] = new TrieNode();
      }
      curNode = curNode.next[c];
    }
    curNode.word = word;
    curNode.isWord = true;
  };

  public search = (word: string) => {
    let curNode: TrieNode = this.root;
    for (let i = 0; i < word.length; i++) {
      curNode = curNode.next[word.charAt(i)];
      if (!curNode) return undefined;
    }
    return curNode.word;
  };

  public display = (node: TrieNode) => {
    console.log(node.word);
    if (node.next) {
      Object.values(node.next).forEach((n: TrieNode) => {
        this.display(n);
      });
    }
  };

  public initialize = () => {
    this.root = new TrieNode();
  };
}
